src/main/scala/rise/autotune/configFileGeneration.scala

```diff
@@ -299,4 +300,35 @@ object configFileGeneration {
     output.toSeq.sortBy(_.name)
   }
 
+  // function to check cycle in parameter dependencies
+  def check_no_cycle(distribution: Map[NatIdentifier,
+    (Set[constraints.Constraint], Set[NatIdentifier])]
+                    ): Boolean = {
+
+    def check_no_cycle_rec(param: NatIdentifier,
+                           dependencies: Set[NatIdentifier],
+                           distribution: Map[NatIdentifier,
+                             (Set[constraints.Constraint], Set[NatIdentifier])],
+                           visited: Set[NatIdentifier]
+                          ): Boolean = {
+
+      dependencies.size match {
+        case 0 => true
+        case _ => dependencies.exists(dependency => {
+          dependency.name.equals(param.name) || visited.contains(dependency)
+        }) match {
+          case true => false
+          case false => dependencies.forall(dependency => {
+            check_no_cycle_rec(param,
+              distribution.apply(dependency)._2,
+              distribution,
+              visited + dependency)
+          })
+        }
+      }
+    }
+
+    distribution.forall(param =>
+      check_no_cycle_rec(param._1, param._2._2, distribution, Set.empty[NatIdentifier]))
```

Comment:
Complexity-wise, I think that what you are doing is roughly `O(V*(V + E))`: for every `V` vertices/parameters, you might visit the entire graph of `V` parameters and `E` edges/dependencies. If you care about performance (and also want to use a time-tested algorithm), you can use standard `O(V + E)` algorithms:
https://www.geeksforgeeks.org/detect-cycle-in-a-graph/
or
https://www.geeksforgeeks.org/detect-cycle-direct-graph-using-colors/